File: src/core/tts/TtsEngine.ts
import Constants, { ExecutionEnvironment } from 'expo-constants';
import { createAudioPlayer, type AudioPlayer } from 'expo-audio';

import type { TtsModelDescriptor } from '../../config/ttsModels';
import { hasSherpaOnnx, tryRequireFs } from '../nativeModules';

export interface SpeakResult {
  synthesisMs: number;
  audioDurationMs: number;
}

/**
 * One loaded native TTS voice, plus its playback.
 *
 * Mirrors SherpaSttBackend's shape on the STT side: a thin wrapper around
 * the native engine's create/use/destroy lifecycle, with exactly one voice
 * resident at a time. `speak()` synthesizes to a WAV file (via the
 * package's own `saveAudioToFile`, not a hand-rolled WAV writer) and plays
 * it with expo-audio — the same audio library the mic side already depends
 * on, so this adds no new audio dependency.
 */
export class TtsEngine {
  private native: any = null;
  private player: AudioPlayer | null = null;
  private tempWavPath: string | null = null;
  private loadedModelId: string | null = null;

  get isReady(): boolean {
    return this.native !== null;
  }

  get loadedFor(): string | null {
    return this.loadedModelId;
  }

  /** Load (or swap to) the voice at `modelPath`. Disposes any previously loaded voice first. */
  async load(model: TtsModelDescriptor, modelPath: string): Promise<void> {
    if (this.loadedModelId === model.id && this.native) return;

    const { createTTS } = requireSherpaTts();
    await this.disposeNative();

    this.native = await createTTS({
      modelPath: { type: 'file', path: modelPath },
      modelType: model.modelType,
      numThreads: 2,
    });
    this.loadedModelId = model.id;
  }

  /**
   * Synthesize `text` and play it. Resolves once playback has actually
   * started (not finished) — callers that need to know when speech ends
   * should use `onFinished`.
   */
  async speak(
    text: string,
    onFinished: () => void,
    onPlaybackError: (message: string) => void
  ): Promise<SpeakResult> {
    if (!this.native) throw new Error('No TTS voice loaded');

    const synthStart = Date.now();
    const audio = await this.native.generateSpeech(text);
    const synthesisMs = Date.now() - synthStart;
    const audioDurationMs = audio.samples.length
      ? (audio.samples.length / audio.sampleRate) * 1000
      : 0;

    const fs = tryRequireFs();
    if (!fs) throw new Error('Filesystem module unavailable in this build.');

    await this.cleanupTempFile();
    const wavPath = `${fs.DocumentDirectoryPath}/tts-playback-${Date.now()}.wav`;
    await saveAudioToFile(audio, wavPath);
    this.tempWavPath = wavPath;

    this.player?.remove();
    const player = createAudioPlayer({ uri: wavPath });
    this.player = player;

    const subscription = player.addListener('playbackStatusUpdate', (status) => {
      if (status.didJustFinish) {
        subscription.remove();
        onFinished();
      }
    });

    try {
      player.play();
    } catch (error) {
      subscription.remove();
      onPlaybackError(messageOf(error));
    }

    return { synthesisMs, audioDurationMs };
  }

  /** Stop playback immediately, without disposing the loaded voice. Used for critical interruption. */
  stopPlayback(): void {
    try {
      this.player?.pause();
    } catch {
      // Best effort.
    }
  }

  async dispose(): Promise<void> {
    await this.disposeNative();
    await this.cleanupTempFile();
  }

  private async disposeNative(): Promise<void> {
    this.player?.remove();
    this.player = null;
    const current = this.native;
    this.native = null;
    this.loadedModelId = null;
    try {
      await current?.destroy?.();
    } catch {
      // Best effort; a failed teardown must not block the next load.
    }
  }

  private async cleanupTempFile(): Promise<void> {
    if (!this.tempWavPath) return;
    const path = this.tempWavPath;
    this.tempWavPath = null;
    try {
      const fs = tryRequireFs();
      if (fs && (await fs.exists(path))) await fs.unlink(path);
    } catch {
      // Non-fatal: a leftover temp WAV costs space, not correctness.
    }
  }
}

function requireSherpaTts(): { createTTS: (options: any) => Promise<any> } {
  if (!hasSherpaOnnx()) {
    throw new Error(
      'Expo Go cannot load react-native-sherpa-onnx (it ships native libraries). ' +
        'Run `npx expo run:android` for a development build to enable speech playback.'
    );
  }
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const mod = require('react-native-sherpa-onnx/tts');
    if (typeof mod?.createTTS !== 'function') {
      throw new Error('createTTS missing from react-native-sherpa-onnx/tts');
    }
    return mod;
  } catch (error) {
    throw new Error(`react-native-sherpa-onnx TTS unavailable: ${String(error)}`);
  }
}

function saveAudioToFile(audio: any, filePath: string): Promise<string> {
  // eslint-disable-next-line @typescript-eslint/no-require-imports
  const mod = require('react-native-sherpa-onnx/tts');
  return mod.saveAudioToFile(audio, filePath);
}

function messageOf(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

File: android-native/app/src/main/java/com/itantra/app/tts/TtsEngine.kt
package com.itantra.app.tts

import android.media.AudioAttributes
import android.media.MediaPlayer
import com.itantra.app.config.TtsModelDescriptor
import com.k2fsa.sherpa.onnx.GeneratedAudio
import com.k2fsa.sherpa.onnx.OfflineTts
import com.k2fsa.sherpa.onnx.OfflineTtsConfig
import com.k2fsa.sherpa.onnx.OfflineTtsModelConfig
import com.k2fsa.sherpa.onnx.OfflineTtsVitsModelConfig
import java.io.File

data class SpeakResult(val synthesisMs: Long, val audioDurationMs: Double)

/**
 * Direct sherpa-onnx TTS engine, adapted from src/core/tts/TtsEngine.ts.
 *
 * Calls com.k2fsa.sherpa.onnx.* directly (same AAR as Phase 2's STT
 * SttEngine), modeled on react-native-sherpa-onnx's SherpaOnnxTtsHelper.kt
 * "vits" branch (node_modules/react-native-sherpa-onnx/android/src/main/
 * java/com/sherpaonnx/SherpaOnnxTtsHelper.kt, buildTtsConfig()'s "vits"
 * case) — the only TTSModelType this app's TTS registry uses (both Piper
 * and MMS voices are VITS models). No React Native TurboModule bridge.
 *
 * The RN wrapper resolves a model directory's file layout via a native C++
 * "model detect" helper that is internal to that package's own JNI glue,
 * not part of the public sherpa-onnx Kotlin API — so it cannot be called
 * directly from here. Instead this scans the directory using the exact
 * layout already documented in src/config/ttsModels.ts and
 * TtsModelManager.ts: a single `*.onnx` file (the VITS model), `tokens.txt`,
 * and — for Piper voices only, which need espeak-ng for text-to-phoneme
 * conversion — a shared `espeak-ng-data/` subdirectory as `dataDir` (left
 * empty for MMS voices, which the source's own comment states do not need
 * it). `lexicon` is left empty for both, matching the source registry
 * (neither packaging documents a lexicon.txt).
 *
 * Playback uses android.media.MediaPlayer in place of expo-audio's
 * AudioPlayer — the same kind of native-Android replacement Phase 3 made
 * for audio capture (AudioRecord replacing expo-audio's stream) — since
 * expo-audio is an Expo/RN library with no place in a from-scratch native
 * Kotlin app.
 */
class TtsEngine {
    private var tts: OfflineTts? = null
    private var loadedModelId: String? = null
    private var player: MediaPlayer? = null
    private var tempWavFile: File? = null

    val isReady: Boolean
        get() = tts != null

    val loadedFor: String?
        get() = loadedModelId

    /** Load (or swap to) the voice at [modelPath]. Disposes any previously loaded voice first. */
    fun load(model: TtsModelDescriptor, modelPath: String) {
        if (loadedModelId == model.id && tts != null) return
        disposeNative()

        val dir = File(modelPath)
        val onnxFile = dir.listFiles()?.firstOrNull { it.name.endsWith(".onnx") }
            ?: throw IllegalStateException("No .onnx model file found in $modelPath")
        val tokensFile = File(dir, "tokens.txt")
        check(tokensFile.exists()) { "Missing tokens.txt in $modelPath" }
        val dataDir = File(dir, "espeak-ng-data")

        val vitsConfig = OfflineTtsVitsModelConfig(
            model = onnxFile.absolutePath,
            lexicon = "",
            tokens = tokensFile.absolutePath,
            dataDir = if (dataDir.exists()) dataDir.absolutePath else "",
        )
        val modelConfig = OfflineTtsModelConfig(
            vits = vitsConfig,
            numThreads = 2,
            debug = false,
            provider = "cpu",
        )
        tts = OfflineTts(config = OfflineTtsConfig(model = modelConfig))
        loadedModelId = model.id
    }

    /**
     * Synthesize [text] and play it via [cacheDir]-backed temp WAV. Returns
     * once playback has actually started (not finished) — callers that need
     * to know when speech ends should use [onFinished].
     */
    fun speak(
        text: String,
        cacheDir: File,
        onFinished: () -> Unit,
        onPlaybackError: (String) -> Unit,
        /**
         * A CRITICAL message plays at the player's full scale.
         *
         * This is only half of "as loud as Android permits" — it removes any
         * per-player attenuation, while [TtsManager] separately raises the
         * media stream itself. Both are needed: a player at 1.0 on a stream
         * turned down to 2/15 is still barely audible.
         */
        critical: Boolean = false,
    ): SpeakResult {
        val engine = checkNotNull(tts) { "No TTS voice loaded" }

        val synthStart = System.currentTimeMillis()
        val audio = engine.generate(text, sid = 0, speed = 1.0f)
        val synthesisMs = System.currentTimeMillis() - synthStart
        val audioDurationMs = if (audio.samples.isNotEmpty()) {
            (audio.samples.size.toDouble() / audio.sampleRate) * 1000.0
        } else {
            0.0
        }

        cleanupTempFile()
        sweepOrphanedWavs(cacheDir)
        val wavFile = File(cacheDir, "tts-playback-${System.currentTimeMillis()}.wav")
        val saved = GeneratedAudio(audio.samples, audio.sampleRate).save(wavFile.absolutePath)
        check(saved) { "Failed to save synthesized audio to ${wavFile.absolutePath}" }
        tempWavFile = wavFile

        player?.release()
        val mp = MediaPlayer()
        player = mp
        try {
            // Declared explicitly rather than left to the default. SPEECH
            // content lets the platform apply speech-appropriate processing,
            // and naming MEDIA usage keeps playback on the same stream the
            // manager requests audio focus for - they have to agree, or the
            // focus request governs a stream the audio is not on.
            mp.setAudioAttributes(
                AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_MEDIA)
                    .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                    .build()
            )
            mp.setVolume(1f, 1f)
            mp.setDataSource(wavFile.absolutePath)
            mp.setOnCompletionListener { onFinished() }
            mp.setOnErrorListener { _, what, extra ->
                onPlaybackError("MediaPlayer error (what=$what, extra=$extra)")
                true
            }
            mp.prepare()
            mp.start()
        } catch (e: Exception) {
            onPlaybackError(e.message ?: "Playback failed to start")
        }

        return SpeakResult(synthesisMs, audioDurationMs)
    }

    /**
     * Stop playback immediately, without disposing the loaded voice. Used for
     * critical interruption.
     *
     * `stop()` rather than `pause()`: this is invoked when a CRITICAL message
     * pre-empts whatever is speaking, and the intent is that the interrupted
     * audio is finished with, not merely suspended mid-word with a decoder
     * still holding the stream. A paused player also stays holding its output
     * until something else releases it, which on the interrupt path is
     * exactly the moment the next voice wants it.
     */
    fun stopPlayback() {
        try {
            player?.stop()
        } catch (e: Exception) {
            // A player already stopped, released, or never started throws
            // rather than no-opping. Nothing to recover here.
        }
    }

    fun dispose() {
        disposeNative()
        cleanupTempFile()
    }

    private fun disposeNative() {
        player?.release()
        player = null
        tts?.release()
        tts = null
        loadedModelId = null
    }

    /**
     * Delete playback WAVs left behind by earlier app sessions.
     *
     * [cleanupTempFile] only knows about the one file this instance is
     * currently tracking, so a process death mid-playback - or simply
     * closing the app - strands its WAV forever. Each is a megabyte or two
     * and they accumulate silently in the cache; four had built up on the
     * test device inside a day.
     */
    private fun sweepOrphanedWavs(cacheDir: File) {
        try {
            val keep = tempWavFile?.name
            cacheDir.listFiles { file ->
                file.isFile && file.name.startsWith("tts-playback-") &&
                    file.name.endsWith(".wav") && file.name != keep
            }?.forEach { it.delete() }
        } catch (e: Exception) {
            // Housekeeping only - never fail a transmission over disk tidiness.
        }
    }

    private fun cleanupTempFile() {
        val file = tempWavFile ?: return
        tempWavFile = null
        try {
            if (file.exists()) file.delete()
        } catch (e: Exception) {
            // Non-fatal: a leftover temp WAV costs space, not correctness.
        }
    }
}
